aQute.libg/src/aQute/lib/memoize/CloseableMemoizingSupplier.java

```diff
@@ -2,74 +2,109 @@
 
+import java.util.concurrent.locks.StampedLock;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 
+/**
+ * The object can exist in one of three states:
+ * <ul>
+ * <li>initial which means neither {@code get} or {@code close} has been called.
+ * From this state, the object can transition directly to either open or
+ * closed.</li>
+ * <li>open which means memoized is non-null with the value from the wrapped
+ * supplier. From this state, the object can transition to closed.</li>
+ * <li>closed which means memoized is null and any value it may have held was
+ * closed. This is a terminal state.</li>
+ * </ul>
+ */
 class CloseableMemoizingSupplier<T extends AutoCloseable> implements CloseableMemoize<T> {
 	private final Supplier<? extends T>	supplier;
-	// @GuardedBy("this")
-	private boolean						closed	= false;
-	private volatile T					memoized;
+	private final StampedLock			lock;
+	private volatile boolean			initial;
+	// @GuardedBy("initial")
+	private T							memoized;
 
 	CloseableMemoizingSupplier(Supplier<? extends T> supplier) {
 		this.supplier = requireNonNull(supplier);
+		lock = new StampedLock();
+		initial = true;
 	}
 
 	@Override
 	public T get() {
-		if (memoized == null) {
-			synchronized (this) {
-				if (closed)
-					throw new IllegalStateException("Already closed");
-				return get0();
+		// read initial _before_ read memoized
+		if (initial) {
+			// critical section: only one at a time
+			final long stamp = lock.writeLock();
+			try {
+				if (initial) {
+					T result = requireNonNull(supplier.get());
+					memoized = result;
+					// write initial _after_ write memoized
+					initial = false;
+					return result;
+				}
+			} finally {
+				lock.unlockWrite(stamp);
 			}
 		}
-		return memoized;
-	}
-
-	private T get0() {
-		if (memoized == null) {
-			memoized = supplier.get();
-			assert memoized != null;
+		T result = memoized;
+		if (result == null) {
+			throw new IllegalStateException("closed");
 		}
-		return memoized;
+		return result;
 	}
 
 	@Override
 	public T peek() {
+		// read initial _before_ read memoized
+		if (initial) {
+			return null;
+		}
 		return memoized;
 	}
 
 	@Override
-	public synchronized boolean isClosed() {
-		return closed;
+	public boolean isClosed() {
+		// read initial _before_ read memoized
+		return !initial && (memoized == null);
 	}
 
 	@Override
 	public void close() throws Exception {
-		T current;
-		synchronized (this) {
-			if (closed)
-				return;
-
-			closed = true;
-
-			if (memoized == null)
-				return;
-
-			current = memoized;
-			memoized = null;
+		if (!isClosed()) {
+			AutoCloseable closeable;
+			// critical section: only one at a time
+			final long stamp = lock.writeLock();
+			try {
+				if (initial) {
+					initial = false;
+					return; // no value to close
+				}
+				closeable = memoized;
+				memoized = null; // mark closed
+				// write initial _after_ write memoized
+				initial = false; // even though it is already false
```

Comment:
The code may be a little non-straight forward but it is designed to only have a single read barrier for the normal case on get. I have review other performance tuned code which does similar work and similar techniques are used to optimize the normal case.

Reply:
it is an interesting puzzle ... my objection is that it does make the code hard to understand. It is a lot clearer now but I've always been pretty hard on simple is better unless proven otherwise. You lately seem to more and more go to small optimizations over readability. 

Just for theory, let me try to come up with something that is more readable and only has one read barrier for the common case.

Did you introduce the StampedLock (first time I run into it...) because you can then have a read lock on the accept? Or is there some other advantage over a synchronized block?

Reply:
> Did you introduce the StampedLock (first time I run into it...) because you can then have a read lock on the accept?

I used it for the read/write lock support which is much simpler then ReentrantReadWriteLock. Some actions (initial get and close) need exclusive (write) locks while accept needs a non-exclusive (read) lock to hold off close. The accept method can now get the write lock if needed and downgrade to the read lock before calling the consumer.

StampedLocks are much faster than ReentrantReadWriteLocks but they are not reentrant (which we did not need) and also have some other limitations (which were not an issue for this usage).

Reply:
I noticed StampedLocks are not reentrant. So if accept calls accept, we have a deadlock. Shouldn't we have a timeout to break this?

Reply:
read locks are non-exclusive, so we wont have any issue with reentrancy on accept. The reentrant code will just get another non-exclusive read lock and then release it when it returns. I confirmed this by tweaking the test case to reentrantly call accept.

```diff
diff --git a/aQute.libg/test/aQute/lib/memoize/MemoizeTest.java b/aQute.libg/test/aQute/lib/memoize/MemoizeTest.java
index 5589a943d..155313cd1 100644
--- a/aQute.libg/test/aQute/lib/memoize/MemoizeTest.java
+++ b/aQute.libg/test/aQute/lib/memoize/MemoizeTest.java
@@ -368,18 +368,18 @@ public class MemoizeTest {
                CountDownLatch consumerReady = new CountDownLatch(multi);
                CountDownLatch consumerSync = new CountDownLatch(1);
                CountDownLatch consumerDone = new CountDownLatch(multi);
-               Consumer<AutoCloseable> consumer = asConsumer(s -> {
-                       consumerReady.countDown();
-                       if (consumerSync.await(20, TimeUnit.SECONDS)) {
-                               consumerDone.countDown();
-                       }
-               });
 
                CountDownLatch threadReady = new CountDownLatch(multi);
                CountDownLatch threadSync = new CountDownLatch(1);
                ExecutorService threadPool = Executors.newFixedThreadPool(multi);
                Supplier<CloseableClass> source = () -> new CloseableClass(count.incrementAndGet());
                try (CloseableMemoize<AutoCloseable> memoized = CloseableMemoize.closeableSupplier(source)) {
+                       Consumer<AutoCloseable> consumer = asConsumer(s -> {
+                               consumerReady.countDown();
+                               if (consumerSync.await(20, TimeUnit.SECONDS)) {
+                                       memoized.accept(x -> consumerDone.countDown());
+                               }
+                       });
                        for (int i = 0; i < multi; i++) {
                                threadPool.execute(asRunnable(() -> {
                                        threadReady.countDown();
```